src/utils/os.c

```diff
@@ -286,6 +286,40 @@ int read_command_output(int fd, process_callback_fn fn, void *ctx) {
   return count;
 }
 
+char **copy_argv(const char *const argv[]) {
+  // calculate the length of argv (without NULL terminator)
+  size_t argc = 0;
+  while (argv[argc] != NULL) {
+    argc++;
+  }
+
+  // calculate the new argv buffer size
+  size_t argv_array_size = (argc + 1) * sizeof(char *);
+  size_t strings_length = 0;
+  for (size_t i = 0; i < argc; i++) {
+    strings_length += (strlen(argv[i]) + 1);
+  }
+
+  // first part of malloc-d data holds argv, second part stores string data
+  char **const argv_copy = (char **)malloc(argv_array_size + strings_length);
+  if (argv_copy == NULL) {
+    log_errno("Failed to malloc %d bytes", argv_array_size + strings_length);
+    return NULL;
+  }
+  char *const argv_string_buffer = &((char *)argv_copy)[argv_array_size];
```

Comment:
Not clear where `argv_copy[0]` is assigned.

Reply:
`argv_copy[0]` is assigned a few lines down: https://github.com/nqminds/edgesec/blob/e6bb700294f822654c55b1e06afc70ec01315441/src/utils/os.c#L314

The first half of the `malloc()`-ed data is filled `char*` (string pointers).
The second half of the `malloc()`-ed data is where the string buffer/contents are.

Essentially, if you copy `const char * argv[] = {"Hello", "World!", NULL};`, the data structure of `argv_copy` looks something like the following:
 
| **Address** |   0-7 |  8-15 | 16-23 |   24-29 |    30-36 |
| ----------- | ----: | ----: | ----: | ------- | -------- |
|  **CType**  | char* | char* | char* |  char[] |   char[] |
|  **Value**  |    24 |    30 |  NULL | "Hello" | "World!" |

I've added some more comments, since I had to make the data structure a bit confusing so that a single `free()` will delete all the data, see https://github.com/nqminds/edgesec/blob/430604f5b03b41a36d87ab265c5247fb711c41b4/src/utils/os.c#L332-L336